### PR TITLE
Add Caffe pose estimation Model

### PR DESCRIPTION
**Pull request by pankaja0285**
Summary of change:
The change refers to the following files
QA/src/evaluatepose.py
QA/src/posepoints.py
QA/src/utils.py
posepy.py

 under the directory, cgm-ml/src/common/reliability/pose-estimation/ 
The script, "evaluatepose.py” for pose estimating the images and saving the pose estimated points as a json file (can also be modified to save as csv file). This script is the start script in the experiment, run as a notebook poseest_notebook.ipynb since we are pose-estimating for a large dataset (in the case that was run was for ~ 110,000 images), has to be run as an experiment pipeline.

Background:
The pretrained CAFFE Pose estimation model is run on (1/6)th of the samples from the "anon_RGB_training" dataset and the results are stored in {experimentname}_pose_points.json". As the json file is huge this could not be uploaded in git.
In order to create the json file, please run the Pose estimation notebook located in the directory. "src/common/reliability/pose_estimation/QA/poseest_notebook.ipynb"
Prerequisites and dependencies:
Following files are required to run the notebook - poseest_notebook.ipynb:-
1.	QA/src/config.py
2.	QA/src/constants.py
3.	QA/src/evaluatepose.py
4.	QA/src/posepoints.py
5.	QA/src/utils.py
6.	commonpy.py
Required library to run evaluatepose.py:
•	pandas
•	PIL
•	numpy
•	cv2
•	tensorflow

Required weights have to be downloaded to a folder called pose right under the main folder pose-estimation from 
https://github.com/legolas123/cv-tricks.com/tree/master/OpenCV/Pose_Estimation
where you can find the weights folder pose
Results of the pose estimation – pose points is saved as a json file. This is given as input to pose analysis task.

**Important**: The pretrained weights need to be downloaded from the following link and has to be in the folder called **pose** under pose-estimation
**sh get_model.sh** from the github folder
https://github.com/legolas123/cv-tricks.com/tree/master/OpenCV/Pose_Estimation
or download the folder pose under the pose-estimation folder and have it under your main pose-estimation folder for running the experiment.

Tests:-
Sample test for fewer images can be found in EDA_poseestimation.ipynb to check pose estimation.
under test blocks at the end of the notebook.

Peer reviewed:-
 Yes
By team mates.

### CODE REVIEW

**Comment by markus-hinsche** [comment]
hi @pankaja0285 Thanks for your contribution. 
- Could you make sure the checks of this pull request are passing?
- Can you add unit tests?

File: src/common/reliability/pose_estimation/QA/src/constants.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from pathlib import Path
+
+REPO_DIR = Path(__file__).parents[3].absolute()
```
Comment (markus-hinsche) [inline src/common/reliability/pose_estimation/QA/src/constants.py:3]: This is not the repository dir, but some child dir of the repository dir

File: src/common/reliability/pose_estimation/posepy.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import cv2 
+import numpy as np
+import pandas as pd
+import time
+import os,sys,inspect
+import matplotlib.pyplot as plt
+from PIL import Image
+from IPython.display import display
+
+def poseEstimationFromModelAndDataset(net, BODY_PARTS, POSE_PAIRS, datasetTypeAndModel,
```
Comment (markus-hinsche) [inline src/common/reliability/pose_estimation/posepy.py:10]: Please use snake_case for all your functions

Comment (markus-hinsche) [inline src/common/reliability/pose_estimation/posepy.py:10]: Please make parameters lower-case

File: src/common/reliability/pose_estimation/posepy.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    # vconcat for combining the titlebox and frameWBorder
+    imframe = cv2.vconcat((titlebox, frameWBorder))
+    # set Pose Estimation file name for imwrite
+    impath = 'PoseEst-' + imageFilename 
+    if (writeImage == True):
+        #create folder if need be
+        if not os.path.exists('output'):
+            os.makedirs('output', mode=0o777, exist_ok=False)
+        # write the file
+        res = cv2.imwrite(f'./output/'+impath, imframe)
```
Comment (markus-hinsche) [inline src/common/reliability/pose_estimation/posepy.py:77]: Please use format strings consistently
```suggestion
        res = cv2.imwrite(f"./output/{impath}", imframe)
```

Look for other occurrences in this PR where `+` can be replaced by a modern format string

File: src/common/reliability/pose_estimation/config.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (markus-hinsche) [inline src/common/reliability/pose_estimation/config.json:1]: Please remove this file, this is not needed here

File: src/common/reliability/pose_estimation/QA/src/constants.py (hunk 1 above)
Comment (pankaja0285) [reply src/common/reliability/pose_estimation/QA/src/constants.py:3]: Regarding your comment about this not being needed, I have to explain why it is needed, its needed for offline run of the experiment notebook namely
poseest_notebook.ipynb
NOTE: I followed the experiment pattern that CGM normally follows.

File: src/common/reliability/pose_estimation/posepy.py (hunk 2 above)
Comment (pankaja0285) [reply src/common/reliability/pose_estimation/posepy.py]: Snake casing has been used for the functions.

Comment (pankaja0285) [reply src/common/reliability/pose_estimation/posepy.py]: Parameters have been made lower case.

File: src/common/reliability/pose_estimation/posepy.py (hunk 3 above)
Comment (pankaja0285) [reply src/common/reliability/pose_estimation/posepy.py]: Consistent format using the modern formatting of string with f"" has been applied.

File: src/common/reliability/pose_estimation/config.json (hunk 4 above)
Comment (pankaja0285) [reply src/common/reliability/pose_estimation/config.json]: Removed the config.json file

File: src/common/reliability/pose_estimation/posepy.py (hunk 2 above)
Comment (markus-hinsche) [reply src/common/reliability/pose_estimation/posepy.py]: I can't see this:
- datasetTypeAndModel should be dataset_type_and_model

Check if this applies to other function names or variable/parameter names

**Comment by markus-hinsche** [comment]
@pankaja0285 Notice that the check in the bottom of the PR are still showing as failing: 

Screenshot:
image

This is due to flake8 issues. Please try to resolve these

**Comment by markus-hinsche** [comment]
@pankaja0285 Can you tell us if you plan to continue working on this?

**Comment by pankaja0285** [comment]
Hi Marcus,  I want to continue to work, but these flakes8 I don't know how to get rid of. If you can help/suggest?I looked up online and someone suggests to do 
rm -rf ~/.cache/pre-commit
But not sure if its ok. Let me know I can check out a file make minor changes and check back in and prior to commit I can run that command.
Thanks,Pankaja Shankar

    On Monday, December 21, 2020, 3:29:58 AM EST, Markus Hinsche <notifications@github.com> wrote:  
 
 


@pankaja0285 Can you tell us if you plan to continue working on this?

**Comment by markus-hinsche** [comment]
@pankaja0285 Maybe @cishwarya or another Omdena team member can help you

**Comment by markus-hinsche** [comment]
This was merged to our codebase. Thanks for the contribution

**Pull request merged**